File: app/javascript/adminterface/lib/inputs/base/pluginish.js
/* global adminterface */
import { toSnakeCase, getObjectValue } from '../../utils'

class Config {
  constructor (configs, options) {
    this.name = configs.name
    this.configs = configs
    this.options = options[toSnakeCase(this.name)]
  }

  attributes () {
    if (!this._enabled()) return {}

    let results

    if (typeof (this.options) === 'object') {
      results = this.options
    } else {
      results = this._defaults()
    }

    return { [`data-aa-${this.name}`]: results }
  }

  _defaults () {
    return this.configs.options || {}
  }

  _enabled () {
    return (
      (!(this.options === false) && this.configs.default === true) ||
        (this.options && this.options === true)
    )
  }
}

class Pluginish {
  constructor (key, options) {
    this.key = key
    this.options = options
    this.plugins = this._plugins()
  }

  enabled (name) {
    return getObjectValue(this.plugins, `data-aa-${name}`)
  }

  _plugins () {
    return Object.assign({}, ...this._configs())
  }

  _configs () {
    const plugins = getObjectValue(adminterface, `meta.components.inputs.${this.key}.js`)

    if (plugins) {
      return plugins.map((plugin) => (new Config(plugin, this.options)).attributes())
    } else {
      return []
    }
  }
}

export default Pluginish

File: lib/adminterface/extensions/inputs/base/pluginish.rb
module Adminterface
  module Extensions
    module Inputs
      module Base
        module Pluginish
          class Config
            attr_reader :name, :configs, :options

            def initialize(configs, options)
              @name = configs[:name]
              @configs = configs
              @options = options[@name.underscore.to_sym]
            end

            def attributes
              return {} unless enabled?

              {"aa-#{name}": options.is_a?(Hash) ? options : defaults}
            end

            def defaults
              @defaults ||= configs[:options] || {}
            end

            def enabled?
              (!options.eql?(false) && configs[:default].eql?(true)) ||
                (options.present? || options.eql?(true))
            end
          end

          def plugins_configs
            plugins&.map { |plugin| Config.new(plugin, options).attributes } || []
          end

          def plugins_options
            {data: plugins_configs.inject(&:merge)}
          end

          def plugin?(name)
            !!plugins_options.dig(:data, "aa-#{name}".to_sym)
          end

          private

          def plugins
            inputs_config.dig(plugins_configs_key, :js)
          end

          def plugins_configs_key
            self.class.name.underscore.split("/").last.gsub("_input", "").to_sym
          end
        end
      end
    end
  end
end
